client/iotcore/client.go

```diff
@@ -0,0 +1,315 @@
+				PolicyDocument: aws.String(policy),
+				PolicyName:     aws.String(policyName(deviceID)),
+			})
+	}
+	var privKey *ecdsa.PrivateKey
+	if err == nil {
+		privKey, err = newPrivateKey()
+	}
+
+	var csr []byte
+	if err == nil {
+		csr, err = newCSR(deviceID, privKey)
+	}
+
+	var respCert *iot.CreateCertificateFromCsrOutput
+	if err == nil {
+		respCert, err = svc.CreateCertificateFromCsrWithContext(ctx,
+			&iot.CreateCertificateFromCsrInput{
+				CertificateSigningRequest: aws.String(string(csr)),
+				SetAsActive:               aws.Bool(device.Status == StatusEnabled),
+			})
+	}
+
+	if err == nil {
+		_, err = svc.AttachPolicyWithContext(ctx,
+			&iot.AttachPolicyInput{
+				PolicyName: respPolicy.PolicyName,
+				Target:     respCert.CertificateArn,
+			})
+	}
+
+	if err == nil {
+		_, err = svc.AttachThingPrincipalWithContext(ctx,
+			&iot.AttachThingPrincipalInput{
+				Principal: respCert.CertificateArn,
+				ThingName: aws.String(deviceID),
+			})
+	}
+
+	var deviceResp *Device
+	if err == nil {
+		deviceResp = &Device{
+			ID:          *resp.ThingId,
+			Name:        *resp.ThingName,
+			Status:      device.Status,
+			PrivateKey:  string(privateKeyToPem(privKey)),
+			Certificate: *respCert.CertificatePem,
+		}
+	}
+	return deviceResp, err
+}
+
+func (c *client) DeleteDevice(ctx context.Context, sess *session.Session, deviceID string) error {
+	svc := iot.New(sess)
+
+	respDescribe, err := svc.DescribeThingWithContext(ctx,
+		&iot.DescribeThingInput{
+			ThingName: aws.String(deviceID),
+		})
+
+	var respListThingPrincipals *iot.ListThingPrincipalsOutput
+	if err == nil {
+		respListThingPrincipals, err = svc.ListThingPrincipalsWithContext(ctx,
+			&iot.ListThingPrincipalsInput{
+				ThingName: aws.String(deviceID),
+			})
+	}
+	if err == nil {
+		for _, principal := range respListThingPrincipals.Principals {
+			_, err := svc.DetachThingPrincipalWithContext(ctx,
+				&iot.DetachThingPrincipalInput{
+					Principal: aws.String(*principal),
+					ThingName: aws.String(deviceID),
+				})
+			var certificateID string
+			if err == nil {
+				parts := strings.SplitAfter(*principal, "/")
+				certificateID = parts[len(parts)-1]
+
+				_, err = svc.UpdateCertificateWithContext(ctx,
+					&iot.UpdateCertificateInput{
+						CertificateId: aws.String(certificateID),
+						NewStatus:     aws.String(iot.CertificateStatusInactive),
+					})
+			}
+			if err == nil {
+				_, err = svc.DeleteCertificateWithContext(ctx,
+					&iot.DeleteCertificateInput{
+						CertificateId: aws.String(certificateID),
+						ForceDelete:   aws.Bool(true),
+					})
+			}
+			if err != nil {
+				break
+			}
+		}
+	}
+
+	if err == nil {
+		_, err = svc.DeleteThingWithContext(ctx,
+			&iot.DeleteThingInput{
+				ThingName:       aws.String(deviceID),
+				ExpectedVersion: aws.Int64(*respDescribe.Version),
+			})
+	}
+	if err != nil {
+		if _, ok := err.(*iot.ResourceNotFoundException); ok {
+			err = ErrDeviceNotFound
+		}
+		return err
+	}
```

Comment:
cant we/should we bail out earlier? I mean: for instance if `DetachThingPrincipalWithContext` returns error above, there is no reason to continue, right? and we can return with error immediately (save a bit of those ifs maybe).

Reply:
Correct; on the other hand, this will lower significantly the test code coverage.
Bringing the err forward is a common pattern in golang.